File: mobile/lib/features/tournaments/domain/dtos/update_tournament_dto.dart
import '../models/tournament_model.dart';

class UpdateTournamentDto {
  final String? name;
  final String? description;
  final DateTime? startDate;
  final DateTime? endDate;
  final List<UpdateCategoryDto>? categories;

  const UpdateTournamentDto({
    this.name,
    this.description,
    this.startDate,
    this.endDate,
    this.categories,
  });

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{};
    if (name != null) json['name'] = name;
    if (description != null) json['description'] = description;
    if (startDate != null) json['startDate'] = startDate!.toIso8601String();
    if (endDate != null) json['endDate'] = endDate!.toIso8601String();
    if (categories != null) {
      json['categories'] = categories!.map((e) => e.toJson()).toList();
    }
    return json;
  }
}

class UpdateCategoryDto {
  final String? id;
  final String? name;
  final CategoryGender? gender;
  final TournamentFormat? format;

  const UpdateCategoryDto({this.id, this.name, this.gender, this.format});

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{};
    if (id != null) json['id'] = id;
    if (name != null) json['name'] = name;
    if (gender != null) json['gender'] = gender!.value;
    if (format != null) json['format'] = format!.value;
    return json;
  }
}

File: backend/src/application/dtos/UpdateTournamentDto.ts
/**
 * DTO para actualizar un torneo existente.
 * Todos los campos son opcionales para permitir actualizaciones parciales.
 */
export interface UpdateTournamentDto {
    name?: string;
    description?: string;
    startDate?: Date;
    endDate?: Date;
    registrationDeadline?: Date;
    location?: string;
    categories?: UpdateCategoryDto[];
}

/**
 * DTO para actualizar una categoría dentro de un torneo.
 */
export interface UpdateCategoryDto {
    /** ID de la categoría a actualizar (opcional para nuevas) */
    id?: string;
    name?: string;
    gender?: 'MALE' | 'FEMALE' | 'MIXED';
    format?: 'SINGLE_ELIMINATION' | 'HYBRID';
    minAge?: number;
    maxAge?: number;
    groupStageConfig?: {
        numberOfGroups: number;
        advancePerGroup: number;
        pointsForWin: number;
        pointsForDraw: number;
        pointsForLoss: number;
        seedingMethod: 'RANKING' | 'RANDOM';
    };
}
